### PR TITLE
allow heuristic based lookahead score functions to use nondemos

### PR DESCRIPTION
**Pull request by tomsilver**

### CODE REVIEW

File: src/approaches/grammar_search_invention_approach.py [hunk 1]
```diff
@@ -656,12 +657,20 @@ def _evaluate_with_operators(self, predicates: FrozenSet[Predicate],
...
         scores = {name: 0.0 for name in self.heuristic_names}
         seen_demos = 0
+        seen_nondemos = 0
+        max_demos = CFG.grammar_search_heuristic_based_max_demos
+        max_nondemos = CFG.grammar_search_heuristic_based_max_nondemos
         for traj, atoms_sequence in pruned_atom_data:
-            if not traj.is_demo:  # we only care about demonstrations
+            # Skip this trajectory if it's not a demo and we only want demos,
+            # or if we've exceeded the budget for demos or nondemos.
+            if (self.demos_only and not traj.is_demo) or (
```
Comment (ronuchit) [inline src/approaches/grammar_search_invention_approach.py:666]: confusing; let's write as 3 separate `if ...: continue`

File: src/approaches/grammar_search_invention_approach.py [hunk 2]
```diff
@@ -647,6 +647,7 @@ class _HeuristicBasedScoreFunction(_OperatorLearningBasedScoreFunction):
     against the demonstrations.
     """
     heuristic_names: Sequence[str]
+    demos_only: bool = field(default=True)
```
Comment (ronuchit) [inline src/approaches/grammar_search_invention_approach.py:650]: are we planning to make a CFG flag for this later?

File: src/settings.py [hunk 3]
```diff
@@ -135,6 +135,7 @@ class GlobalSettings:
     grammar_search_score_function = "hff_lookahead_depth0"
     grammar_search_heuristic_based_weight = 10.
     grammar_search_heuristic_based_max_demos = 5
+    grammar_search_heuristic_based_max_nondemos = 50
```
Comment (ronuchit) [inline src/settings.py:138]: is there no hope of it ever running with the full 500, even in cover?

File: src/approaches/grammar_search_invention_approach.py (hunk 1 above)
Comment (tomsilver) [reply src/approaches/grammar_search_invention_approach.py:666]: I think 2 is clear, so I'm going to do that

File: src/approaches/grammar_search_invention_approach.py (hunk 2 above)
Comment (tomsilver) [reply src/approaches/grammar_search_invention_approach.py:650]: I'm planning to make a different score function that uses `demos_only=False` later

Comment (ronuchit) [reply src/approaches/grammar_search_invention_approach.py:650]: that would be a subclass of this one?

File: src/settings.py (hunk 3 above)
Comment (tomsilver) [reply src/settings.py:138]: we can make it env-dependent later if you want, I imagine 500 would be fine for cover, but 50 is also seemingly fine for cover

File: src/approaches/grammar_search_invention_approach.py (hunk 2 above)
Comment (tomsilver) [reply src/approaches/grammar_search_invention_approach.py:650]: yeah well this would be a mixin, like we do for the existing lookahead score functions

**Review by ronuchit — APPROVED** [review]
lgtm

**Pull request merged**